### PR TITLE
Feature/lp 284 patch email

### PR DESCRIPTION
**Pull request by lduranteau**
### JIRA link
https://companieshouse.atlassian.net/browse/LP-284

### Change description
add patch endpoint to save email in db

### Work checklist

* [ ] Bug fix
* [x] New feature
* [ ] Breaking change
* [ ] Infrastructure change

### Pull request checklist

* [x] I have added unit tests for new code that I have added
* [ ] I have added/updated functional tests where appropriate
* [x] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__
### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 1]
```diff
@@ -62,4 +70,39 @@ public ResponseEntity<Object> createPartnership(
...
+    ) throws JsonProcessingException {
+
+        String transactionId = transaction.getId();
+        HashMap<String, Object> logMap = new HashMap<>();
+        logMap.put(URL_PARAM_TRANSACTION_ID, transactionId);
+
+        try {
+            String type = (String) body.get("type");
+            var dataType = DataType.valueOf(type.toUpperCase());
+            final Map<String, Object> data = extractData(body);
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:90]: could you catch the JsonProcessingException in this try and handle it the same as the ServiceException?

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java [hunk 2]
```diff
@@ -69,6 +71,27 @@ public String createLimitedPartnership(Transaction transaction,
...
+
+        if (limitedPartnershipSubmissionDao.isEmpty()) {
+            throw new ServiceException(String.format(
+                    "Submission with id %s not found", submissionId));
+        }
+
+        if (type == DataType.EMAIL) {
+            DataDao dataDao = limitedPartnershipSubmissionDao.get().getData();
+
+            String email = (String) data.get("email");
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:88]: can "email" be replaced with DataType.EMAIL.getDescription() ?

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:88]: or better, type.getDescription()

**Comment by markpit** [comment]
looks like 1 code smell still present

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 1 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:90]: need more explanation about what you want here

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java (hunk 2 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/service/LimitedPartnershipService.java:88]: this line get the value send by the web not a type

**Comment by markpit** [comment]
the spec/schema.json file needs to be updated as well

**Comment by lduranteau** [comment]
> looks like 1 code smell still present

don't understand why sonarqube prefers "var" instead of the type.
I had to change a lot

**Comment by lduranteau** [comment]
> > looks like 1 code smell still present
> 
> don't understand why sonarqube prefers "var" instead of the type. I had to change a lot

changed to var

**Comment by lduranteau** [comment]
> the spec/schema.json file needs to be updated as well

spec updated

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java (hunk 1 above)
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:90]: updated

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 3]
```diff
@@ -62,4 +70,39 @@ public ResponseEntity<Object> createPartnership(
...
     }
+
+    @PatchMapping("/{" + URL_PARAM_SUBMISSION_ID + "}")
+    public ResponseEntity<Object> updatePartnership(
+            @RequestAttribute(TRANSACTION_KEY) Transaction transaction,
+            @PathVariable(URL_PARAM_SUBMISSION_ID) String submissionId,
+            @RequestBody Map<String, Object> body,
+            @RequestHeader(value = ERIC_REQUEST_ID_KEY) String requestId,
+            @RequestHeader(value = ERIC_IDENTITY) String userId
+    ) throws JsonProcessingException {
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java:81]: do you still need the throws JsonProcessingException ? you should be able to remove it now we are catching it in the method.

Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: removed

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by markpit** [comment]
LGTM should probably get 1 more review

**Review by dlloyd1-ch — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java [hunk 4]
```diff
@@ -62,4 +70,39 @@ public ResponseEntity<Object> createPartnership(
...
+        try {
+            String type = (String) body.get("type");
+            var dataType = DataType.valueOf(type.toUpperCase());
+            final Map<String, Object> data = extractData(body);
+
+            limitedPartnershipService.updateLimitedPartnership(submissionId, dataType, data);
+
+            return new ResponseEntity<>(HttpStatus.OK);
+        } catch (JsonProcessingException | ServiceException e) {
+            ApiLogger.errorContext(requestId, "Error creating Limited Partnership submission", e, logMap);
```
[the remark this answers is not in the document]
Comment (lduranteau) [reply src/main/java/uk/gov/companieshouse/limitedpartnershipsapi/controller/PartnershipController.java]: updated

**Review by mwestacott — APPROVED** [review]

**Pull request merged**